modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -1180,6 +1199,88 @@
     (for [i (range (apply max (map count posts)))]
       (into {} (map #(get % i)) posts))))
 
+(mu/defn- order-by->$sort :- [:map-of ::lib.schema.common/non-blank-string [:enum -1 1]]
+  [order-by :- [:sequential ::mbql.s/OrderBy]]
+  (into
+   (ordered-map/ordered-map)
+   (for [[direction field] order-by]
+     [(->lvalue field) (case direction
+                         :asc   1
+                         :desc -1)])))
+
+(defn- generate-window-output-clause
+  "Takes a pair of [output-name input-name] and generates an output clause suitable for
+  including in a `$setWindowFields` output block."
+  [[output-name input-name]]
+  {output-name
+   {$sum input-name
+    "window" {"documents" ["unbounded" "current"]}}})
+
+(defn- sort-lookup
+  "Generates a lookup string for a particular field"
+  [id name]
+  (if (id name)
+    (str "_id." name)
+    name))
+
+(defn- make-window-sort
+  "Converts a `$sort` body to something that can be used in a `sortBy` clause in a
+  `$setWindowFields` stage."
+  [id pairs]
+  (when-let [pair-seq (seq pairs)]
+    (into (ordered-map/ordered-map)
+          (map (fn [[name dir]] [(sort-lookup id name) dir]))
+          pair-seq)))
+
+(defn- get-window-sort-and-partitions
+  "Calculates the appropriate sort and partition fields for a `$setWindowFields` stage."
+  [id breakouts order-by]
+  (let [finest-temporal-index
+        (qp.util.transformations.nest-breakouts/finest-temporal-breakout-index breakouts 2)
+
+        sort-index (or finest-temporal-index
+                       (dec (count breakouts)))
+        sort-name (first (nth (seq id) sort-index))
+        default-sort {(sort-lookup id sort-name) 1}
+        user-sort (when order-by
+                    (binding [*field-mappings*
+                              (merge *field-mappings*
+                                     (into {} (map (juxt identity field-alias)) breakouts))]
+                      (order-by->$sort order-by)))
+        sort-expr (cond
+                    ;; if there is only one breakout, always use the user's sort order
+                    (= (count id) 1)
+                    (or (make-window-sort id user-sort) default-sort)
+
+                    ;; if we don't have a temporal breakout, sort by the last breakout, but
+                    ;; use the user's sort direction if specified
+                    (not finest-temporal-index)
+                    (or (->> user-sort
+                             (filter #(= sort-name (first %)))
+                             (make-window-sort id))
+                        default-sort)
+
+                    :else default-sort)
```

Comment:
This might be more neatly expressed as a big `or` with clauses like `(when-not finest-temporal-index ...)` in it. That's usually better than cond when there's several branches and they all end with "or use `some-default`".

Reply:
Sure, makes sense